File: hdar/archives/hdar_3/hdar/lifecycle/state_machine.py
"""Lifecycle state machine — P0 addition #1.

One authoritative state machine governing capsule lifecycle transitions.
A capsule may seal only after the state machine confirms quiescence and
effect reconciliation.

States:
    DORMANT → ACQUIRING_LEASE → MATERIALIZING → VERIFYING_INPUT
    → RUNNING → QUIESCING → SEALING → DESTROYING → DORMANT

Failure states:
    QUARANTINED, DEGRADED, UNKNOWN_EFFECT, LEASE_LOST,
    RESTORE_REJECTED, DESTRUCTION_UNCONFIRMED
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Callable, Dict, List, Optional


class AgentState(Enum):
    DORMANT = auto()
    ACQUIRING_LEASE = auto()
    MATERIALIZING = auto()
    VERIFYING_INPUT = auto()
    RUNNING = auto()
    QUIESCING = auto()
    SEALING = auto()
    DESTROYING = auto()
    # failure states
    QUARANTINED = auto()
    DEGRADED = auto()
    UNKNOWN_EFFECT = auto()
    LEASE_LOST = auto()
    RESTORE_REJECTED = auto()
    DESTRUCTION_UNCONFIRMED = auto()


# Valid forward transitions
TRANSITIONS: Dict[AgentState, List[AgentState]] = {
    AgentState.DORMANT: [AgentState.ACQUIRING_LEASE],
    AgentState.ACQUIRING_LEASE: [AgentState.MATERIALIZING, AgentState.LEASE_LOST, AgentState.DORMANT],
    AgentState.MATERIALIZING: [AgentState.VERIFYING_INPUT, AgentState.RESTORE_REJECTED, AgentState.QUARANTINED],
    AgentState.VERIFYING_INPUT: [AgentState.RUNNING, AgentState.RESTORE_REJECTED, AgentState.QUARANTINED],
    AgentState.RUNNING: [AgentState.QUIESCING, AgentState.UNKNOWN_EFFECT, AgentState.LEASE_LOST, AgentState.QUARANTINED],
    AgentState.QUIESCING: [AgentState.SEALED if False else AgentState.SEALING, AgentState.UNKNOWN_EFFECT],
    AgentState.SEALED if False else AgentState.SEALING: [AgentState.DESTROYING, AgentState.QUARANTINED],
    AgentState.DESTROYING: [AgentState.DORMANT, AgentState.DESTRUCTION_UNCONFIRMED],
    # failure states can transition to DORMANT after resolution
    AgentState.QUARANTINED: [AgentState.DORMANT],
    AgentState.DEGRADED: [AgentState.RUNNING, AgentState.DORMANT],
    AgentState.UNKNOWN_EFFECT: [AgentState.QUIESCING, AgentState.QUARANTINED],
    AgentState.LEASE_LOST: [AgentState.DORMANT],
    AgentState.RESTORE_REJECTED: [AgentState.DORMANT],
    AgentState.DESTRUCTION_UNCONFIRMED: [AgentState.DORMANT, AgentState.QUARANTINED],
}

# Fix the SEALING key (workaround for the conditional above)
TRANSITIONS[AgentState.QUIESCING] = [AgentState.SEALING, AgentState.UNKNOWN_EFFECT]
TRANSITIONS[AgentState.SEALING] = [AgentState.DESTROYING, AgentState.QUARANTINED]


@dataclass
class StateTransition:
    from_state: AgentState
    to_state: AgentState
    timestamp: float
    reason: str
    metadata: Dict = field(default_factory=dict)


class LifecycleStateMachine:
    """Authoritative state machine for one agent's runtime lifecycle.

    Enforces that transitions only happen along valid edges. Records
    every transition for audit. A capsule may seal only from QUIESCING
    or SEALING states.
    """

    def __init__(self, agent_id: str):
        self.agent_id = agent_id
        self.state = AgentState.DORMANT
        self.history: List[StateTransition] = []
        self._transition_time = time.time()

    def transition(self, to_state: AgentState, reason: str = "", metadata: Optional[dict] = None) -> bool:
        """Attempt a state transition. Returns True if valid."""
        allowed = TRANSITIONS.get(self.state, [])
        if to_state not in allowed:
            return False

        record = StateTransition(
            from_state=self.state,
            to_state=to_state,
            timestamp=time.time(),
            reason=reason,
            metadata=metadata or {},
        )
        self.history.append(record)
        self.state = to_state
        self._transition_time = record.timestamp
        return True

    def can_seal(self) -> bool:
        """A capsule may seal only from QUIESCING or SEALING."""
        return self.state in (AgentState.QUIESCING, AgentState.SEALING)

    def is_running(self) -> bool:
        return self.state == AgentState.RUNNING

    def is_dormant(self) -> bool:
        return self.state == AgentState.DORMANT

    def is_failure(self) -> bool:
        return self.state in (
            AgentState.QUARANTINED,
            AgentState.DEGRADED,
            AgentState.UNKNOWN_EFFECT,
            AgentState.LEASE_LOST,
            AgentState.RESTORE_REJECTED,
            AgentState.DESTRUCTION_UNCONFIRMED,
        )

    def time_in_state(self) -> float:
        return time.time() - self._transition_time

    def to_dict(self) -> dict:
        return {
            "agent_id": self.agent_id,
            "state": self.state.name,
            "transitions": [
                {
                    "from": t.from_state.name,
                    "to": t.to_state.name,
                    "timestamp": t.timestamp,
                    "reason": t.reason,
                    "metadata": t.metadata,
                }
                for t in self.history
            ],
        }

    @classmethod
    def from_dict(cls, d: dict) -> "LifecycleStateMachine":
        sm = cls(d["agent_id"])
        sm.state = AgentState[d["state"]]
        for t in d.get("transitions", []):
            sm.history.append(StateTransition(
                from_state=AgentState[t["from"]],
                to_state=AgentState[t["to"]],
                timestamp=t["timestamp"],
                reason=t["reason"],
                metadata=t.get("metadata", {}),
            ))
        return sm

File: hdar/hdar_core/cpp/src/state_machine.cpp
#include "hdar/state_machine.hpp"
#include <set>
#include <map>

namespace hdar {

std::string state_name(AgentState s) {
    switch (s) {
        case AgentState::DORMANT: return "DORMANT";
        case AgentState::ACQUIRING_LEASE: return "ACQUIRING_LEASE";
        case AgentState::MATERIALIZING: return "MATERIALIZING";
        case AgentState::VERIFYING_INPUT: return "VERIFYING_INPUT";
        case AgentState::RUNNING: return "RUNNING";
        case AgentState::QUIESCING: return "QUIESCING";
        case AgentState::SEALING: return "SEALING";
        case AgentState::DESTROYING: return "DESTROYING";
        case AgentState::QUARANTINED: return "QUARANTINED";
        case AgentState::DEGRADED: return "DEGRADED";
        case AgentState::UNKNOWN_EFFECT: return "UNKNOWN_EFFECT";
        case AgentState::LEASE_LOST: return "LEASE_LOST";
        case AgentState::RESTORE_REJECTED: return "RESTORE_REJECTED";
        case AgentState::DESTRUCTION_UNCONFIRMED: return "DESTRUCTION_UNCONFIRMED";
    }
    return "UNKNOWN";
}

static std::map<AgentState, std::set<AgentState>> build_transitions() {
    return {
        {AgentState::DORMANT, {AgentState::ACQUIRING_LEASE}},
        {AgentState::ACQUIRING_LEASE, {AgentState::MATERIALIZING, AgentState::LEASE_LOST, AgentState::DORMANT}},
        {AgentState::MATERIALIZING, {AgentState::VERIFYING_INPUT, AgentState::RESTORE_REJECTED, AgentState::QUARANTINED}},
        {AgentState::VERIFYING_INPUT, {AgentState::RUNNING, AgentState::RESTORE_REJECTED, AgentState::QUARANTINED}},
        {AgentState::RUNNING, {AgentState::QUIESCING, AgentState::UNKNOWN_EFFECT, AgentState::LEASE_LOST, AgentState::QUARANTINED}},
        {AgentState::QUIESCING, {AgentState::SEALING, AgentState::UNKNOWN_EFFECT}},
        {AgentState::SEALING, {AgentState::DESTROYING, AgentState::QUARANTINED}},
        {AgentState::DESTROYING, {AgentState::DORMANT, AgentState::DESTRUCTION_UNCONFIRMED}},
        {AgentState::QUARANTINED, {AgentState::DORMANT}},
        {AgentState::DEGRADED, {AgentState::RUNNING, AgentState::DORMANT}},
        {AgentState::UNKNOWN_EFFECT, {AgentState::QUIESCING, AgentState::QUARANTINED}},
        {AgentState::LEASE_LOST, {AgentState::DORMANT}},
        {AgentState::RESTORE_REJECTED, {AgentState::DORMANT}},
        {AgentState::DESTRUCTION_UNCONFIRMED, {AgentState::DORMANT, AgentState::QUARANTINED}},
    };
}

bool LifecycleStateMachine::is_valid_transition(AgentState from, AgentState to) {
    static auto transitions = build_transitions();
    auto it = transitions.find(from);
    if (it == transitions.end()) return false;
    return it->second.count(to) > 0;
}

LifecycleStateMachine::LifecycleStateMachine(const std::string& aid)
    : agent_id_(aid), transition_time_(epoch_seconds()) {}

bool LifecycleStateMachine::transition(AgentState to, const std::string& reason,
                                        const JsonValue& metadata) {
    if (!is_valid_transition(state_, to))
        return false;

    StateTransition t;
    t.from_state = state_;
    t.to_state = to;
    t.timestamp = epoch_seconds();
    t.reason = reason;
    t.metadata = metadata;
    history_.push_back(std::move(t));
    state_ = to;
    transition_time_ = epoch_seconds();
    return true;
}

bool LifecycleStateMachine::can_seal() const {
    return state_ == AgentState::QUIESCING || state_ == AgentState::SEALING;
}

bool LifecycleStateMachine::is_failure() const {
    return state_ == AgentState::QUARANTINED ||
           state_ == AgentState::DEGRADED ||
           state_ == AgentState::UNKNOWN_EFFECT ||
           state_ == AgentState::LEASE_LOST ||
           state_ == AgentState::RESTORE_REJECTED ||
           state_ == AgentState::DESTRUCTION_UNCONFIRMED;
}

JsonValue LifecycleStateMachine::to_json() const {
    JsonValue v = JsonValue::object();
    v["agent_id"] = JsonValue::string(agent_id_);
    v["state"] = JsonValue::string(state_name(state_));

    JsonValue trans_arr = JsonValue::array();
    for (const auto& t : history_) {
        JsonValue to = JsonValue::object();
        to["from"] = JsonValue::string(state_name(t.from_state));
        to["to"] = JsonValue::string(state_name(t.to_state));
        to["timestamp"] = JsonValue::number(t.timestamp);
        to["reason"] = JsonValue::string(t.reason);
        to["metadata"] = t.metadata;
        trans_arr.push_back(std::move(to));
    }
    v["transitions"] = std::move(trans_arr);
    return v;
}

} // namespace hdar
